src/GovUk.Education.ExploreEducationStatistics.Common/Config/SensitiveDataTelemetryProcessor.cs

```diff
@@ -0,0 +1,160 @@
+using Microsoft.ApplicationInsights.DataContracts;
+using Microsoft.ApplicationInsights.Extensibility;
+
+namespace GovUk.Education.ExploreEducationStatistics.Common.Config
+{
+    /// <summary>
+    /// Tries to filter out any potentially sensitive data from telemetry
+    /// before it is sent to Application Insights e.g. query string parameters
+    /// containing things like access tokens or codes.
+    /// </summary>
+    public class SensitiveDataTelemetryProcessor : ITelemetryProcessor
+    {
+        private const string RedactedValue = "__redacted__";
+
+        private readonly HashSet<string> _redactedTokens = new()
+        {
+            "code",
+            "email",
+            "phone",
+            "token",
+            "passphrase",
+            "password"
+        };
+
+        private readonly char[] _tokenDelimiters =
+        {
+            '_', '-', ':', '.', '|', ' ', '[', ']', '+'
+        };
+
+        private readonly Regex _casingRegex = new("([A-Z][a-z]*|[0-9])", RegexOptions.Compiled);
+
+        private ITelemetryProcessor Next { get; }
+
+        public SensitiveDataTelemetryProcessor(ITelemetryProcessor next)
+        {
+            Next = next;
+        }
+
+        public void Process(ITelemetry item)
+        {
+            switch (item)
+            {
+                case RequestTelemetry telemetry:
+                {
+                    telemetry.Url = FilterUrl(telemetry.Url);
+                    break;
+                }
+                case PageViewTelemetry telemetry:
+                {
+                    telemetry.Url = FilterUrl(telemetry.Url);
+
+                    if (telemetry.Properties.ContainsKey("refUri"))
+                    {
+                        telemetry.Properties["refUri"] = FilterUriQuery(telemetry.Properties["refUri"]);
+                    }
+
+                    break;
+                }
+                case DependencyTelemetry telemetry:
+                {
+                    HandleDependencyTelemetry(telemetry);
+                    break;
+                }
+            }
+
+            Next.Process(item);
+        }
+
+        private void HandleDependencyTelemetry(DependencyTelemetry telemetry)
+        {
+            if (telemetry.Type != "Http")
+            {
+                return;
+            }
+
+            telemetry.Name = FilterUriQuery(telemetry.Name);
+            telemetry.Data = FilterUriQuery(telemetry.Data);
+        }
+
+        private Uri? FilterUrl(Uri? uri)
+        {
+            if (uri is null)
+            {
+                return uri;
+            }
+
+            var filteredUri = FilterUriQuery(uri.AbsoluteUri);
+
+            return filteredUri is null ? uri : new Uri(filteredUri);
+        }
+
+        private string? FilterUriQuery(string? uri)
+        {
+            if (uri.IsNullOrEmpty())
+            {
+                return uri;
+            }
+
+            var pathAndQuery = uri.Split('?', 2);
```

Comment:
Surprised there isn't a less manual way of getting a list of query params from a URI, but I've looked and can't see anything immediately obvious.

Reply:
Oh, or is the point that you still want to try to redact invalid URLs?

Reply:
Yep that's why. Originally, I was doing some parsing of the URLs, but it would have potentially let through malformed URL structures. Dumbing the algorithm down _should_ make it a little more robust for these types of cases.